torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -537,6 +571,16 @@ void TensorPipeAgent::sendCompletedResponseMessage(
           VLOG(1) << "RPC agent for " << workerInfo_.name_
                   << " done sending response to request #" << messageId
                   << " to " << pipe->getRemoteName();
+
+          // use a dedicated thread to synchronize CUDA streams before
+          // destructing them.
+          // NB: when we add a dedicated stream pool later, this prevents
+          // returning the stream to the pool too early.
+          threadPool_.run([streams{std::move(streams)}]() mutable {
```

Comment:
I thought the latest version of our design called for not synchronizing streams (except on the caller side, before marking the future complete; and even that would only be a temporary measure, until we introduce the CUDAFuture). That's because synchronizing streams doesn't change the API behavior of the TensorPipe agent (because remote functions can enqueue on the streams regardless of whether there's already something in it); it may only affect performance, and even on that matter we're still not sure if it actually helps. On the other hand, by synchronizing the CUDA streams like this, we're occupying one thread of the pool, which will negatively affect the performance by reducing the parallelism of remote functions. I'd prefer to first see some benchmarks before we sacrifice CPU perf for some unverified CUDA perf.

Reply:
> it may only affect performance

Are we sure that it won't cause deadlock when using nested RPC if we don't wait for streams?

Reply:
> Are we sure that it won't cause deadlock when using nested RPC if we don't wait for streams?

This may not be a convincing argument, but my "gut" tells me we should be fine.

That's because I think it's veeery hard to create a deadlock in CUDA, if one "plays by the rules".

Let me try to be a bit more formal. (This is going to be an interesting exercise)

Let's start by assuming that the our CPU threads, when they deal with non-CUDA stuff, do _not_ deadlock. We have the guarantee that the CUDA API is thread-safe, meaning that even if two threads work on the same CUDA objects without explicit synchronization (mutexes and condition variables), CUDA will ensure that their operations will be serialized in some total order. This is an important property, which will come in handy in a second.

Let's now look at what I think are the "idiomatic" ways to perform synchronization between CUDA streams: `cudaEventRecord` and `cudaStreamWaitEvent`. These are the asynchronous functions, contrary to `cudaEventSynchronize` and `cudaStreamSynchronize` (which I consider less idiomatic). The way EventRecord+StreamWaitEvent works is somewhat odd (took me a while to figure out, and was the topic of lengthy meetings between me and @beauby). In particular, it is very different from the CPU primitives for synchronization (mutexes, condition variables, futures, ...). This is how I understand them: streams and events, in CUDA, are (mainly) a host-side API construct to allow the developer to create a dependency graph between CUDA ops (kernel launches, memcpys, ...). When I imagine a "dependency graph" I always picture it as arrows. Streams allow to create "chains" of arrows between their ops. And events are a way to add extra arrows between objects in different streams. The way an event works is that it stores the "tail" of an arrow. When one creates an event, it starts by being empty, which is as if the tail of the arrow was at -infty. Once an event is recorded onto some stream (call it "stream A"), the "internal storage" of the event is updated so that the tail of its arrow is now placed at the point in stream A where the event was recorded (i.e., the current tip of the stream). When we call StreamWaitEvent, on that event, on _another_ stream (call it "stream B"), what happens is that a full arrow is drawn onto the dependency graph: the tail of the arrow is the one stored in the event's "internal storage", the head of the arrow is the point at which stream B currently is (its current tip). What this means is that the next CUDA op that is enqueued on stream B will have a dependency on the ops that were present in stream A when the event was recorded.

So far so good (hopefully), but the interesting part comes from the fact that the event's internal state is _mutable_, which means we can update the "arrow tail". The arrows that have already been fully drawn (tail + head) won't change when we modify an event. But all the new arrows we'll draw (i.e., the future calls to StreamWaitEvent) will use the new "arrow tail" position. What this means is that the way arrows are drawn between streams depends on the order in which operations are invoked _on the CPU_. This is where the property I mentioned at the beginning comes to the rescue: we have a total order for the operations invoked on the CPU!

I have a math background so I'll introduce some formulas. Let `evA`, `evB`, `evC`, ... be our events. Let `evX.record[i]` be the "arrow tail" stored by the `i`-th call to EventRecord (`i = 0, 1, ...`). Let `evX.wait[j]` be the "arrow head" drawn by the `j`-th call to StreamWaitEvent (`j = 0, 1, ...`). A full arrow drawn using `evX` thus goes from `evX.record[i]` to `evX.wait[j]`, for some values of `i` and `j`. (The values of `i` and `j` could be different, because one can wait multiple times on the same record, or not wait at all and immediately re-record). The core property though is:
```
cpu_time(evX.record[i]) < cpu_time(evX.wait[j])
```

I now want to prove that in CUDA it's impossible to have a "lock inversion" (a very common type of deadlock). A lock inversion, in CPU-land, consist in two events, A and B, with one thread waiting on A to be set and, afterwards, setting B, and another thread first waiting for B and then setting A. This will block because of basically a dependency cycle: setting A must happen after finishing waiting for B, which must happen after setting B, which must happen after finishing waiting for A, which must happen after setting A.

In CUDA this cannot happen. We want to show that it's impossible to have two events, `evX` and `evY`, such that, on one stream, `evX.wait[j1]` is enqueued before `evY.record[i2]` and, on another stream, `evY.wait[j2]` is enqueued before `evX.record[i1]`. This can only happen if `cpu_time(evX.wait[j1]) < cpu_time(evY.record[i2])` and `cpu_time(evY.wait[j2]) < cpu_time(evX.record[i1])`. Using the above property, though, we also know that `cpu_time(evX.record[i1]) < cpu_time(evX.wait[j1])`, `cpu_time(evY.record[i2]) < cpu_time(evY.wait[j2])`. All these inequalities cannot be true at the same time, we've reached a contradiction, which concludes our proof, hooray!

I think other types of deadlocks can also be proven to be impossible using the same technique.

To trace back our steps, what were our assumptions? Basically, two: that we didn't have deadlocks in CPU threads to start with, and that we limited ourselves to those two "idiomatic" CUDA calls. What happens if we introduce other CUDA ops to the picture? And is there a way that CUDA could introduce deadlocks between CPU threads?

The vast majority of CUDA ops are asynchronous: they return immediately, without waiting for anything. The three notable exceptions are `cudaDeviceSynchronize`, `cudaStreamSynchronize` and `cudaEventSynchronize`. (Other API calls are also blocking if `CUDA_LAUNCH_BLOCKING` is set, but that's a debug option so let's ignore it for now). So if we do _not_ use these three calls CUDA shouldn't be able to introduce any CPU deadlocks, and we're free to use all other CUDA calls.

Using one of these three synchronization functions, however, I believe, is dangerous. I do not have an example in mind where they could cause a CPU deadlock, but since they do hold up the CPU thread they are run from, I consider them to be more at risk. This is one reason why I try as much as possible to shy away from them and use the async "idiomatic" CUDA ops exclusively.

(This comment became sooo much longer than I had originally planned, and I didn't reread because it's 9pm on Friday and I want to call it a day; sorry if this ended up to be an unintelligible rambling, I guess I'll realize that on Monday morning ;) )

Reply:
That's a great explanation! I am convinced by the lock inversion proof. Thanks for putting this together. 

One followup question regarding the behavior of `pipe->write`. IIRC, it will be the callee that launches the `cudaMemcpy`. In TensorPipe's implementation, how does the caller know that the callee is done with the `cudaMemcpy`? Does the caller let the current stream to wait for the event from callee? also cc @beauby

Reply:
Reading through this again, maybe that entire "essay" could have been summed up as: there cannot be any deadlock among CUDA ops which is due to a cycle in the dependency graph because the arrows in the graph are always directed in the same order as the CPU operations that scheduled them, and those API calls are non-blocking and thread-safe and thus form a total order.

One more thing I wanted to clarify: we said that deadlocks are impossible in CUDA, and yet NCCL sometimes deadlocks, how can that be? One silent assumption we made was that every "node" in the dependency graph (namely every kernel and every memcpy), once it's launched it is able to run on its own, independently, and complete within some fixed finite amount of time. Therefore the entire execution graph will eventually run and complete. This is true for all "normal" kernels but NCCL has a "hacky" approach where its I/O kernels communicate with a CPU thread and block on a busy wait. Such a kernel is thus able to stall and hold up the execution of everything that comes after it in the execution graph. In case of a bug, this can cause a deadlock.

Personally, I think that's an "abuse" of the CUDA design (a necessary one, but still an abuse) and thus it's not PyTorch's duty to be prepared to handle it, but it should be NCCL to put in the effort to hide these implementation details and present a CUDA-compliant experience. I'm saying that because TensorPipe will also eventually have to implement such an I/O kernel and I see it as our responsibility to make it behave well.

> In TensorPipe's implementation, how does the caller know that the callee is done with the cudaMemcpy? Does the caller let the current stream to wait for the event from callee?

Correct, TensorPipe will add the necessary events to ensure that the cudaMemcpyAsync is only launched once both streams are ready, and that both streams are "released" (and can continue with the later work) only after the memcpy has finished. By "both streams" I mean both the sender's and the receiver's.

Reply:
addressed in latest version. It now directly uses `CUDAFuture`